substrate/tendermint/machine/src/lib.rs

```diff
@@ -0,0 +1,620 @@
+          panic!("honest node had invalid behavior");
+        }
+
+        match res {
+          Ok(None) => (),
+          Ok(Some(block)) => {
+            let mut validators = vec![];
+            let mut sigs = vec![];
+            // Get all precommits for this round
+            for (validator, msgs) in &self.block.log.log[&msg.round] {
+              if let Some(Data::Precommit(Some((id, sig)))) = msgs.get(&Step::Precommit) {
+                // If this precommit was for this block, include it
+                if id == &block.id() {
+                  validators.push(*validator);
+                  sigs.push(sig.clone());
+                }
+              }
+            }
+
+            let commit = Commit {
+              end_time: self.block.end_time[&msg.round].canonical(),
+              validators,
+              signature: N::SignatureScheme::aggregate(&sigs),
+            };
+            debug_assert!(self.network.verify_commit(block.id(), &commit));
+
+            let proposal = self.network.add_block(block, commit).await;
+            self.reset(msg.round, proposal).await;
+          }
+          Err(TendermintError::Malicious(validator)) => {
+            self.slash(validator).await;
+          }
+          Err(TendermintError::Temporal) => (),
+        }
+
+        if broadcast {
+          let sig = self.signer.sign(&msg.encode()).await;
+          self.network.broadcast(SignedMessage { msg, sig }).await;
+        }
+      }
+    }
+  }
+
+  // Returns Ok(true) if this was a Precommit which had its signature validated
+  // Returns Ok(false) if it wasn't a Precommit or the signature wasn't validated yet
+  // Returns Err if the signature was invalid
+  fn verify_precommit_signature(
+    &self,
+    sender: N::ValidatorId,
+    round: RoundNumber,
+    data: &DataFor<N>,
+  ) -> Result<bool, TendermintError<N::ValidatorId>> {
+    Ok(if let Data::Precommit(Some((id, sig))) = data {
+      // Also verify the end_time of the commit
+      // Only perform this verification if we already have the end_time
+      // Else, there's a DoS where we receive a precommit for some round infinitely in the future
+      // which forces to calculate every end time
+      if let Some(end_time) = self.block.end_time.get(&round) {
+        if !self.validators.verify(sender, &commit_msg(end_time.canonical(), id.as_ref()), sig) {
+          Err(TendermintError::Malicious(sender))?;
+        }
+        true
+      } else {
+        false
+      }
+    } else {
+      false
+    })
+  }
+
+  async fn message(
+    &mut self,
+    msg: MessageFor<N>,
+  ) -> Result<Option<N::Block>, TendermintError<N::ValidatorId>> {
+    if msg.number != self.block.number {
+      Err(TendermintError::Temporal)?;
+    }
+
+    // If this is a precommit, verify its signature
+    self.verify_precommit_signature(msg.sender, msg.round, &msg.data)?;
+
+    // Only let the proposer propose
+    if matches!(msg.data, Data::Proposal(..)) &&
```

Comment:
Because this check is only executed if this is a Proposal. We could use a match statement, yet that's bad practice when you only care about a single branch.

Reply:
Makes it easier to read...

Reply:
It's actually not endorsed by Rust and I believe triggers clippy for a single line match.